File: go/src/problem_003.go
/*
 * The prime factors of 13195 are 5, 7, 13 and 29.
 * 
 * What is the largest prime factor of the number 600851475143?
 */

package main

import "fmt"
import "math"

// var n = 600851475143 uint64

func divisors_of(n uint64) (divisors []uint64) {
  divisors = make([]uint64, 0)
  max := uint64(math.Ceil(math.Sqrt(float64(n))))

  for i := uint64(1); i <= max; i++ {
    if n % i == 0 {
      divisors = append(divisors, i)
    }
  }

  return
}

func is_prime(n uint64) bool {
  for i := uint64(2); i < n; i++ {
    if n % i == 0 {
      return false
    }
  }
  return true
}

func prime_divisors_of(n uint64) (ps []uint64) {
  ps = make([]uint64, 0)
  divs := divisors_of(n)
  for _, v := range divs {
    if is_prime(v) {
      ps = append(ps, v)
    }
  }
  return
}

func main() {
  ps := prime_divisors_of(600851475143)
  fmt.Println(ps[len(ps) - 1])
}

File: c/problem_003.c
/*
 * Problem 3
 *
 * The prime factors of 13195 are 5, 7, 13 and 29.
 *
 * What is the largest prime factor of the number 600,851,475,143?
 */

// TODO: This is not finished yet (obvs).

#include <euler.h>

const uint64_t n = 600851475143;
/* const uint64_t n = 13195; */

uint64_t largest_prime_factor(uint64_t x) {
  mpz_t * prime_sieve = sieve_primes(x);

  for(uint64_t i = x; i >= 2; --i) {
    if(!mpz_tstbit(*prime_sieve, i)) continue;

    if(x % i == 0) return i;
  }
  return x;
}

int main(int argc, char * argv[]) {
  printf("%lu\n", largest_prime_factor(n));
}
